### ISSUE TITLE
-inf value when applying log to p(j | x, y<i)

### ISSUE THREAD

**Issue by greeksharifa** [backprop07/Self-Certainty#6]
Thanks for your great work!

I'm applying this self-centainty method to my code. However, for Qwen-2.5, many of tokens have exactly or near zero probability, so log p(...) would have the -inf value.

How do I solve this problem?

**Comment by backprop07** [comment]
Thanks for the problem. The issue is almost always insufficient precision in logit processing. We’ve only seen it when logits are in float16; switching to bfloat16 or float32 before softmax eliminated the problem on every model we tested, including the Qwen 2.5 series.

**Comment by greeksharifa** [comment]
Thanks for your reply. but we use bfloat16 like the following code snippet:
```python
from transformers import Qwen2_5_VLForConditionalGeneration, AutoProcessor

self.model = Qwen2_5_VLForConditionalGeneration.from_pretrained(
    model_id,
    torch_dtype=torch.bfloat16,
    low_cpu_mem_usage=True,
    attn_implementation="flash_attention_2" if self.cfg.model_cfg.use_flash_attn_2 else None,
    device_map=self.cfg.model_cfg.device_map,
    cache_dir=cache_dir,
    # do_sample=False,
)
```

example tensor:
```bash
(Pdb) torch.log(valid_prob_all_vocab)                                                                                                                                                                                                                                                  
tensor([[-18.4231, -15.0169, -13.2669,  ..., -24.2141, -24.2141, -24.2141],                                                                                                                                                                                                            
        [-17.7026, -16.6401, -17.2026,  ..., -21.9526, -21.9526, -21.9526]],                                                                                                                                                                                                           
       device='cuda:0')                                                                                                                                                                                                                                                                
(Pdb) torch.log(valid_prob_all_vocab).sum()                                                                                                                                                                                                                                            
tensor(-inf, device='cuda:0')  
```

**Comment by backprop07** [comment]
Are any values in `torch.log(valid_prob_all_vocab)` showing `-inf`?

* **If yes**, it likely means that `valid_prob_all_vocab` contains zeros, and taking the log of zero results in `-inf`. In this case, you can try using automatic mixed precision to ensure numerical stability during calculations. For example:

```python
with torch.autocast(device_type="cuda", dtype=torch.bfloat16):
    logits = model(input_ids)  # forward pass
    log_probs = torch.nn.functional.log_softmax(logits, dim=-1).sum(dim=-1)
```

* **If not**, the issue may be due to summing over too many values, causing numerical overflow. To mitigate this, consider:

  * Increasing the floating-point range temporarily (e.g., using a higher-precision dtype like `float64`).
  * Normalizing (e.g., dividing) values before summation to reduce magnitude, then rescaling if needed.
